Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs

```diff
@@ -0,0 +1,258 @@
+            ProjectInfo projectInfo;
+            if (_projectInfo.TryGetValue(project, out projectInfo)) {
+                var analysis = project.GetAnalyzer().GetAnalysisEntryFromPath(path);
+                TestContainer container;
+                if (analysis != null && projectInfo._containers.TryGetValue(analysis, out container)) {
+                    return container;
+                }
+            }
+
+            return null;
+        }
+
+        private static IEnumerable<IVsProject> EnumerateLoadedProjects(IVsSolution solution) {
+            var guid = new Guid(PythonConstants.ProjectFactoryGuid);
+            IEnumHierarchies hierarchies;
+            ErrorHandler.ThrowOnFailure((solution.GetProjectEnum(
+                (uint)(__VSENUMPROJFLAGS.EPF_MATCHTYPE | __VSENUMPROJFLAGS.EPF_LOADEDINSOLUTION),
+                ref guid,
+                out hierarchies)));
+            IVsHierarchy[] hierarchy = new IVsHierarchy[1];
+            uint fetched;
+            while (ErrorHandler.Succeeded(hierarchies.Next(1, hierarchy, out fetched)) && fetched == 1) {
+                var project = hierarchy[0] as IVsProject;
+                if (project != null) {
+                    yield return project;
+                }
+            }
+        }
+
+        public event EventHandler TestContainersUpdated;
+
+        private void OnProjectLoaded(object sender, ProjectEventArgs e) {
+            if (e.Project != null) {
+                var pyProj = e.Project.GetPythonProject();
+                if (pyProj != null) {
+                    var analyzer = pyProj.GetAnalyzer();
+                    var projectInfo = new ProjectInfo(this, pyProj);
+                    pyProj.InterpreterDbChanged += projectInfo.DatabaseChanged;
+                    _projectInfo[pyProj] = projectInfo;
+
+                    foreach (var file in analyzer.LoadedFiles) {
+                        if (file.Value.IsAnalyzed) {
+                            projectInfo.AnalysisComplete(this, new AnalysisCompleteEventArgs(file.Value));
+                        }
+                    }
+                }
+            }
+
+            TestContainersUpdated?.Invoke(this, EventArgs.Empty);
+        }
+
+        class ProjectInfo {
+            private readonly PythonProjectNode _project;
+            private readonly TestContainerDiscoverer _discoverer;
+            public readonly Dictionary<AnalysisEntry, TestContainer> _containers;
+
+            public ProjectInfo(TestContainerDiscoverer discoverer, PythonProjectNode project) {
+                _project = project;
+                _discoverer = discoverer;
+                _containers = new Dictionary<AnalysisEntry, TestContainer>();
+
+                project.ProjectAnalyzerChanged += ProjectAnalyzerChanged;
+                HookAnalysisComplete();
+            }
+
+            private void HookAnalysisComplete() {
+                _project.GetAnalyzer().AnalysisComplete += AnalysisComplete;
```

Comment:
What if the analyzer changes?

---
In reply to: [67420732](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67420732) [](ancestors = 67420732,67384909)